### PR TITLE
Opcode `SHR`

### PR DESCRIPTION
**Pull request by silathdiir**
spec: https://github.com/appliedzkp/zkevm-specs/pull/195

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 1]
```diff
@@ -896,3 +901,263 @@ impl<F: Field> MulAddWordsGadget<F> {
...
+            .collect::<Result<Vec<_>, _>>()?;
+        self.a64s_hi
+            .iter()
+            .zip(a64s_hi.iter())
+            .map(|(cell, val)| cell.assign(region, offset, Some(F::from_u128(*val))))
+            .collect::<Result<Vec<_>, _>>()?;
+        let shf_div64 = shf_div64 as u64;
+        let shf_mod64 = shf_mod64 as u64;
+        self.shf_div64
+            .assign(region, offset, Some(F::from(shf_div64)))?;
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:1143]:
```suggestion
            .assign(region, offset, Some(F::from(shf_div64 as u64)))?;
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 2]
```diff
@@ -896,3 +901,263 @@ impl<F: Field> MulAddWordsGadget<F> {
...
+        self.shf_mod64
+            .assign(region, offset, Some(F::from(shf_mod64)))?;
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:1145]:
```suggestion
            .assign(region, offset, Some(F::from(shf_mod64 as u64)))?;
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 3]
```diff
@@ -896,3 +901,263 @@ impl<F: Field> MulAddWordsGadget<F> {
...
+            .zip(a64s_lo.iter())
+            .map(|(cell, val)| cell.assign(region, offset, Some(F::from_u128(*val))))
+            .collect::<Result<Vec<_>, _>>()?;
+        self.a64s_hi
+            .iter()
+            .zip(a64s_hi.iter())
+            .map(|(cell, val)| cell.assign(region, offset, Some(F::from_u128(*val))))
+            .collect::<Result<Vec<_>, _>>()?;
+        let shf_div64 = shf_div64 as u64;
+        let shf_mod64 = shf_mod64 as u64;
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:1141]: we can remove them

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 4]
```diff
@@ -896,3 +901,263 @@ impl<F: Field> MulAddWordsGadget<F> {
...
+            cb.require_equal(
+                "a64s[idx] == a64s_lo[idx] + a64s_hi[idx] * p_lo",
+                a64s[idx].expr(),
+                a64s_lo[idx].expr() + a64s_hi[idx].expr() * p_lo.expr(),
+            );
+        }
+
+        // a64s_lo[idx] < p_lo
+        let a64s_lo_lt_p_lo =
+            array_init(|idx| LtGadget::construct(cb, a64s_lo[idx].expr(), p_lo.expr()));
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:980]: need to add extra constraint for `a64s_lo_lt_p_lo`. Constructing `LtGadget` doesn't check the `a64s_lo[idx] < p_lo`.

**Review by icemelon — COMMENTED** [review]
after addressing these comments, you can send pr to upstream

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs (hunk 4 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs]: Fixed to add `cb.require_equal` as below:
```
        let a64s_lo_lt_p_lo = array_init(|idx| {
            let lt = LtGadget::construct(cb, a64s_lo[idx].expr(), p_lo.expr());
            cb.require_equal("a64s_lo[idx] < p_lo", lt.expr(), 1.expr());
            lt
        });
```

**Comment by silathdiir** [comment]
Moved to https://github.com/appliedzkp/zkevm-circuits/pull/519

**Pull request closed without merging**